llvm/lib/TargetParser/AArch64TargetParser.cpp

```diff
@@ -48,6 +48,19 @@ std::optional<AArch64::ArchInfo> AArch64::ArchInfo::findBySubArch(StringRef SubA
   return {};
 }
 
+unsigned AArch64::getFMVPriority(ArrayRef<StringRef> Features) {
+  constexpr unsigned MaxFMVPriority = 1000;
+  unsigned Priority = 0;
+  unsigned NumFeatures = 0;
+  for (StringRef Feature : Features) {
+    if (auto Ext = parseFMVExtension(Feature)) {
+      Priority = std::max(Priority, Ext->Priority);
+      NumFeatures++;
+    }
+  }
+  return Priority + MaxFMVPriority * NumFeatures;
```

Comment:
I don't understand the algorithm here. This seems like sorts resolvees first by how many conditions they've got, and _then_ by the highest priority of each in that group. But what we actually need is a total order on two sets. Maybe the API should be shaped more like a comparison operator on feature sets, rather than a serialized index constructed from members's priorities?

Reply:
(and yeah, I see that this is just a pure refactor, so it's not "new", but still...)

Reply:
You're right, that's what it does. The more specific a version is (more features), then higher its priority. If it's a tie, the version with the highest priority feature wins. Indeed this is NFC. Changing it is ABI break, which doesn't bother me personally. I think since FMV is not yet widely used, we should be free to make sensible changes to the specification.

I am not sure what the original idea was. Perhaps it makes sense a feature that depends on other features to have higher priorirty than them? We could use the runtime bitmasks for this as I tried to do here https://github.com/llvm/llvm-project/pull/87939/files#r1842027067. On a tangent, we (Arm) are discussing about potentially detecting dependent features at runtime, but that's for another time - slightly relevant though.

Reply:
I'm not sure it does that correctly: `simd+dotprod` is weighted higher than `sme` with this algorithm purely because it has more features.

Reply:
I am also not bothered by breaking ABI here, for the record.

Reply:
> I'm not sure it does that correctly: I think simd+dotprod is weighted higher than sme with this algorithm purely because it has more features.

Exactly. `Priority(simd+dotprod) > Priority(sme) > Priority(simd) + Priority(dotprod)`. That's intentional according to the current spec.

Reply:
We should probably clarify and/or fix the spec, both orders seem like valid interpretations to me, as there isn't a definition for how to calculate the priority when a version has more than one feature:

> 6.4.3 Selection [^](https://arm-software.github.io/acle/main/acle.html#markdown-toc-selection)
> 
> The following rules shall be followed by all implementations:
> 
> 1. Implementation of the selection algorithm is platform dependent, where with platform means CPU/Vendor/OS as in the target triplet.
> 1. The selection is permanent for the lifetime of the process.
> 1. Only those versions could be considered where all dependent features are available.
>
> Rules of version selection are in order:
> 
> 1. Select the most specific version else
> 1. select the version with the highest priority else
> 1. "default" is selected if no other versions are suitable.

should we ping the gcc folks and get their read on a) how it works for them now, and b) how it should work going forward?

Reply:
> Select the most specific version else

Admittedly vague, but I believe "most specific" means the one with the most features here